_projects/automate_vulture_whitelist.md

```diff
@@ -0,0 +1,78 @@
+---
+name: "Automate Whitelist creation for vulture"
+collaborating_projects:
+  - coala
+  - vulture
+desc: "Automatically create whitelists and detect false positives for vulture."
+developers_involved:
+  - RJ722
+mentors:
+  - AbdealiJK
+  - jendrikseipp
+  - The-Compiler
+difficulty: medium
+status: in_progress
+initiatives:
+  - GSoC
+issues: []
+markdown: automate_vulture_whitelist.md
+requirements:
+  - "Familiarity with vulture and how whitelists work."
+tags:
+  - Vulture
+  - AST
+  - coverage
+---
+
+The aim of this project is to automatically create whitelists (to be utilized
+by [vulture](https://github.com/jendrikseipp/vulture)) for projects using sip
+files to communicate between C++ level API and Python programs, most
+significant of them being PyQt projects.
+
+#### IDEAS
+
+* Create a script which takes sip files as input and spits a whitelist with all
+  the virtual functions defined in the sip files. (Add Documentation and tests)
+* Use that script for generating a whitelist for popular projects using PyQt
+  (like [qutebrowser](https://github.com/qutebrowser/qutebrowser)).
+* Allow passing coverage.xml to vulture via `--make-whitelist` and let it find
+  the list of false-positive functions. (Add Documentation and tests as well)
+* Write a whitelist generator for shiboken, the binding
+  generator used for PySide2
+
+#### MILESTONES
+
+##### COMMUNITY BONDING
+* Discussion on whether to implement a new Tracer or to directly consume `xml`
+  output from coverage.py
+* Get familiar with coverage.xml
+* Take up small tasks from
+  [TODO](https://github.com/jendrikseipp/vulture/tree/master/TODO.rst)
+* [Follow up milestone](https://gitlab.com/coala/GSoC/GSoC-2018/milestones/28)
+
+##### CODING PHASE 1
+* The script for creating a whitelist from sip files should be implemented.
+* It should reside in it's own repository, with CI enabled for testing on both
```

Comment:
is this a coala org repo, or is vulture going to become an org?  or under your own user space?

Presumably these whitelists could also be used by coala's python bears, and other python tools, which are currently unaware of voodoo provided by frameworks and C wrappers.

Reply:
> is this a coala org repo, or is vulture going to become an org? or under your own user space?

I'd very much like input from @AbdealiJK and @jendrikseipp.

> Presumably these whitelists could also be used by coala's python bears, and other python tools, which are currently unaware of voodoo provided by frameworks and C wrappers.

I cannot see a simple use case where other coala bears or python tools can use these whitelists. Do you have any ideas in specific @jayvdb?

Reply:
The whitelist is probably useful for any project using PyQt (or PySide2), but not for other projects.

Reply:
I'd call the repo RJ722/vulture-whitelist-generators for now. I think it'll become clearer where to put this later.

Reply:
> I'd call the repo RJ722/vulture-whitelist-generators for now. I think it'll become clearer where to put this later.

Hehe, Thanks for the honor @jendrikseipp!

Reply:
> I cannot see a simple use case where other coala bears or python tools

All of them could use it, for the same reasons as vulture.  Maybe not simple.

e.g. see our pyflakes project cEP.

Reply:
> Whitelists are just Python files containing a reference to an object which was reported as unused by vulture

This synopsis is precisely why the problem should not be though of in terms of only vulture.

The work being done is not identifying references to an object or method which was reported as unused by vulture.

The work being done is identifying references to ast elements that not unused, *but* no Python AST-only based algorithms can detect without inspecting external definition files and/or coverage data, or at least these means of gathering the data is 'better' than trying to obtain it from Python AST of dependencies.

It should be a generic tool (library) which provides the information which was determined by using sip/shiboken/coverage, and then a vulture specific use of that information which pushes that information into the whitelist format which vulture wants.



pyflakes has a whitelist mechanism, the [`PYFLAKES_BUILTINS`](https://github.com/PyCQA/pyflakes/blob/a13dbe6/pyflakes/checker.py#L478) envvar, which could be populated by data obtained from your tools, especially the information obtained from coverage.xml.

This same library could also be utilised by a new bear using [pyflakes AST](https://github.com/coala/cEPs/blob/master/cEP-0025.md), which is @MacBox7's project.

Bears written using the antlr AST will likely have the same problems.

Reply:
in order to achieve this we would be needing an inverse mapping i.e given the line and column provide us with `pyflakes node`. We could then get a list of unused code from some tools (Vulture or pyflakes itself) and cross check it with coverage report to generate whitelist.

Reply:
this project would only provide a list of symbols found in scopes *not* part of the pyflakes ast analysis - see project for where it plans to get them from. It isnt necessary to know line or col : the primary value is in knowing about extra names which are valid.

These symbol lists are only useful to augment the inbuilt data. See pyflakes envvar mentioned in my last comment. Specifically to prevent false positives being report.

building an unused code detector bear is not a useful use case as that would be reinventing the vulture wheel.

Reply:
I understand this. But the coverage report returned by [coverage.py](https://coverage.readthedocs.io/en/coverage-4.5.1/) only returns lineno and not the list of symbols. Thus an inverse mapping is required to return nodes corresponding to the line no. These symbols can then be supplied to envvar.

Reply:
The workflow would be something like this:
- run Vulture on the project
- whatever functions or methods vulture reports as unused, check whether or not that particular line number(s) are hit or not in the `coverage.xml`'s output.

An inverse mapping would be required if you are going the opposite way, from `coverage`'s output to the `ast`.